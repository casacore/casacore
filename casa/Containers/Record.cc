//# Record.cc: A hierarchical collection of named fields of various types
//# Copyright (C) 1995,1996,1997,1998,1999,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casacore/casa/Containers/Record.h>
#include <casacore/casa/Containers/RecordDesc.h>
#include <casacore/casa/IO/AipsIO.h>
#include <casacore/casa/Arrays/IPosition.h>
#include <casacore/casa/Exceptions/Error.h>
#include <casacore/casa/Utilities/Assert.h>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

Record::Record()
: RecordInterface (),
  rep_p    (new RecordRep),
  parent_p (0)
{}

Record::Record (RecordType type,
		CheckFieldFunction* func, const void* checkArgument)
: RecordInterface (type, func, checkArgument),
  rep_p    (new RecordRep),
  parent_p (0)
{}
	
Record::Record (const RecordDesc& description, RecordType type,
		CheckFieldFunction* func, const void* checkArgument)
: RecordInterface (type, func, checkArgument),
  rep_p    (new RecordRep (description)),
  parent_p (0)
{}

// When description is empty, Record structure is variable.
Record::Record (RecordRep* parent, const RecordDesc& description)
: RecordInterface (description.nfields()==0 ? Variable : Fixed, 0, 0),
  rep_p    (new RecordRep (description)),
  parent_p (parent)
{}

Record::Record (RecordRep* parent, RecordType type)
: RecordInterface (type, 0, 0),
  rep_p    (new RecordRep),
  parent_p (parent)
{}

Record::Record (const Record& other)
: RecordInterface (other),
  rep_p    (other.rep_p),
  parent_p (other.parent_p)
{}

Record::Record (const RecordInterface& other)
: RecordInterface (other),
  rep_p    (new RecordRep (other.description())),
  parent_p (0)
{
    uInt n = other.nfields();
    const RecordDesc& desc = description();
    for (uInt i=0; i<n; i++) {
	DataType dtype = desc.type(i);
	if (dtype == TpRecord) {
	    const RecordInterface& subrec = *((RecordInterface*)
                                              (other.get_pointer (i, dtype)));
	    defineRecord (i, Record (subrec),
			  subrec.isFixed()  ?  Fixed : Variable);
	}else{
	    rep_p->copyDataField (dtype, i, other.get_pointer (i, dtype));
	}
    }
}

Record& Record::operator= (const Record& other)
{
    // Assignment is only possible when the Record is empty or
    // when their layout match or when the Record is non-fixed.
    // When non-fixed or empty, we simply replace the representation.
    // Otherwise we replace all values (in which case we do not need
    // to replace the pointers in RecordFieldPtr's).
    if (this != &other) {
	if (! isFixed()  ||  nfields() == 0) {
	    rep_p = other.rep_p;
	}else{
	    AlwaysAssert (conform (other), AipsError);
	    rwRef().copyData (other.ref());
	}
    }
    return *this;
}

Record::~Record()
{}

RecordInterface* Record::clone() const
{
    return new Record (*this);
}

void Record::assign (const RecordInterface& that)
{
    *this = that;
}

void Record::print (ostream& os, Int maxNrValues, const String& indent) const
{
    rep_p.ref().print (os, maxNrValues, indent);
}

void Record::makeUnique()
{
    rwRef();
}

RecordRep& Record::rwRef()
{
    return rep_p.rwRef();
}

const String& Record::comment (const RecordFieldId& id) const
{
    Int whichField = idToNumber (id);
    return ref().comment (whichField);
}
void Record::setComment (const RecordFieldId& id, const String& comment)
{
    Int whichField = idToNumber (id);
    rwRef().setComment (whichField, comment);
}

RecordDesc Record::getDescription() const
{
    return ref().description();
}

void Record::restructure (const RecordDesc& newDescription, Bool recursive)
{
    // Restructure is not possible for fixed records.
    throwIfFixed();
    // Restructuring means that all RecordFieldPtr's get invalid.
    rwRef().restructure (newDescription, recursive);
}

uInt Record::nfields() const
{
    return description().nfields();
}
Int Record::fieldNumber (const String& fieldName) const
{
    return description().fieldNumber (fieldName);
}
DataType Record::type (Int whichField) const
{
    return description().type (whichField);
}

void Record::removeField (const RecordFieldId& id)
{
    throwIfFixed();
    Int whichField = idToNumber (id);
    rwRef().removeField (whichField);
}

void Record::renameField (const String& newName, const RecordFieldId& id)
{
    rwRef().renameField (newName, idToNumber(id));
}


void Record::addDataField (const String& name, DataType type,
			   const IPosition& shape, Bool fixedShape,
			   const void* value)
{
    rwRef().addDataField (name, type, shape, fixedShape, value);
}

void Record::defineDataField (Int whichField, DataType type,
			      const void* value)
{
    rwRef().defineDataField (whichField, type, value);
}

void* Record::get_pointer (Int whichField, DataType type) const
{
    return ref().get_pointer (whichField, type);
}
void* Record::get_pointer (Int whichField, DataType type,
			   const String& recordType) const
{
    return ref().get_pointer (whichField, type, recordType);
}

void Record::defineRecord (const RecordFieldId& id,
			   const RecordInterface& value, RecordType type)
{
    defineRecord (id, Record (value), type);
}
void Record::defineRecord (const RecordFieldId& id,
			   const Record& value, RecordType type)
{
    Int whichField = newIdToNumber (id);
    if (whichField < 0) {
	throwIfFixed();
	String name;
	if (id.byName()) {
	    name = id.fieldName();
	}else{
	    name = description().makeName (id.fieldNumber());
	}
	checkName (name, TpRecord);
	rwRef().addField (name, value, type);
    }else{
	rwRef().defineDataField (whichField, TpRecord, &value);
	Record& subrec = *(Record*)get_pointer (whichField, TpRecord);
	subrec.recordType() = type;
    }
}


const RecordInterface& Record::asRecord (const RecordFieldId& id) const
{
    return subRecord (id);
}
RecordInterface& Record::asrwRecord (const RecordFieldId& id)
{
    return rwSubRecord (id);
}
const Record& Record::subRecord (const RecordFieldId& id) const
{
    Int whichField = idToNumber (id);
    return *(const Record*)get_pointer (whichField, TpRecord);
}
Record& Record::rwSubRecord (const RecordFieldId& id)
{
    Int whichField = idToNumber (id);
    rwRef();
    return *(Record*)get_pointer (whichField, TpRecord);
}    


void Record::mergeField (const Record& other, const RecordFieldId& id,
			 DuplicatesFlag flag)
{
    throwIfFixed();
    Int whichField = other.idToNumber (id);
    rwRef().mergeField (other.ref(), whichField, flag);
}

void Record::merge (const Record& other, DuplicatesFlag flag)
{
    AlwaysAssert (this != &other, AipsError);
    throwIfFixed();
    rwRef().merge (other.ref(), flag);
}


void Record::putRecord (AipsIO& os) const
{
    ref().putRecord (os, recordType());
}
void Record::getRecord (AipsIO& os)
{
    // Get is only possible when the Record is empty or when
    // the Record is non-fixed.
    AlwaysAssert ((! isFixed()  ||  nfields() == 0), AipsError);
    // Reading the record type back means casting it from an int
    // to the correct type.
    Int type;
    rwRef().getRecord (os, type);
    recordType() = (RecordInterface::RecordType)type;
}

} //# NAMESPACE CASACORE - END

