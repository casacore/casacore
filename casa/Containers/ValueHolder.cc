//# ValueHolder.cc: A holder object for the standard AIPS++ data
//# Copyright (C) 2005
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//#
//# $Id$


#include <casa/Containers/ValueHolder.h>
#include <casa/Containers/Record.h>
#include <casa/Arrays/Array.h>
#include <casa/Arrays/ArrayMath.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions/Error.h>

namespace casa { //# NAMESPACE CASA - BEGIN


ValueHolder::ValueHolder (Bool value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (uChar value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (Short value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (uShort value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (Int value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (uInt value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (Float value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (Double value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Complex& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const DComplex& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Char* value)
  : itsRep (new ValueHolderRep(String(value)))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const String& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Bool>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<uChar>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Short>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<uShort>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Int>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<uInt>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Float>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Double>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<Complex>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<DComplex>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Array<String>& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (const Record& value)
  : itsRep (new ValueHolderRep(value))
{
  itsRep->link();
}

ValueHolder::ValueHolder (uInt ndim, Bool dummy)
  : itsRep (new ValueHolderRep(ndim, dummy))
{
  itsRep->link();
}


ValueHolder::ValueHolder (const ValueHolder& that)
  : itsRep (that.itsRep)
{
  if (itsRep) itsRep->link();
}

ValueHolder& ValueHolder::operator= (const ValueHolder& that)
{
  if (this != &that) {
    ValueHolderRep::unlink (itsRep);
    itsRep = that.itsRep;
    if (itsRep) itsRep->link();
  }
  return *this;
}


} //# NAMESPACE CASA - END
