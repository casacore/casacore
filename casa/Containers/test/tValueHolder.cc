//# tValueHolder.cc: Test the ValueHolder class
//# Copyright (C) 2009
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/Containers/ValueHolder.h>
#include <casa/Containers/Record.h>
#include <casa/Utilities/Assert.h>
#include <casa/Arrays/Vector.h>
#include <casa/Arrays/ArrayLogical.h>
#include <casa/Arrays/ArrayMath.h>

using namespace casa;

void doBool (Bool v)
{
  ValueHolder vh(v);
  AlwaysAssertExit (!vh.isNull());
  AlwaysAssertExit (vh.dataType() == TpBool);
  AlwaysAssertExit (vh.asBool() == v);
  Vector<Bool> vec = vh.asArrayBool();
  AlwaysAssertExit (vec.size() == 1  &&  vec.data()[0] == v);
  cout << vh << endl;
  Record rec;
  vh.toRecord (rec, "a");
  ValueHolder vhc;
  AlwaysAssertExit (vhc.isNull());
  vhc = ValueHolder::fromRecord (rec, "a");
  AlwaysAssertExit (!vhc.isNull());
  AlwaysAssertExit (vhc.dataType() == TpBool);
  Bool vc;
  vhc.getValue (vc);
  AlwaysAssertExit (vc == v);
  cout << vh << ' ' << vhc << endl;
}

template<typename T, typename U> void doPos(T v, U, DataType dt)
{
  ValueHolder vh(v);
  AlwaysAssertExit (!vh.isNull());
  AlwaysAssertExit (vh.dataType() == dt);
  AlwaysAssertExit (vh.asBool());
  AlwaysAssertExit (vh.asuChar() == uChar(v));
  AlwaysAssertExit (vh.asShort() == Short(v));
  AlwaysAssertExit (vh.asuShort() == uShort(v));
  AlwaysAssertExit (vh.asInt() == Int(v));
  AlwaysAssertExit (vh.asuInt() == uInt(v));
  AlwaysAssertExit (vh.asInt64() == Int64(v));
  AlwaysAssertExit (vh.asFloat() == float(v));
  AlwaysAssertExit (vh.asDouble() == double(v));
  AlwaysAssertExit (vh.asComplex() == Complex(float(v),0));
  AlwaysAssertExit (vh.asDComplex() == DComplex(double(v),0));
  Vector<Bool> vecbool = vh.asArrayBool();
  AlwaysAssertExit (vecbool.size() == 1  &&  vecbool.data()[0] == True);
  Vector<uChar> vecuChar = vh.asArrayuChar();
  AlwaysAssertExit (vecuChar.size() == 1  &&  vecuChar.data()[0] == uChar(v));
  Vector<Short> vecShort = vh.asArrayShort();
  AlwaysAssertExit (vecShort.size() == 1  &&  vecShort.data()[0] == Short(v));
  Vector<uShort> vecUShort = vh.asArrayuShort();
  AlwaysAssertExit (vecUShort.size() == 1  &&  vecUShort.data()[0] == uShort(v));
  Vector<Int> vecInt = vh.asArrayInt();
  AlwaysAssertExit (vecInt.size() == 1  &&  vecInt.data()[0] == Int(v));
  Vector<uInt> vecuInt = vh.asArrayuInt();
  AlwaysAssertExit (vecuInt.size() == 1  &&  vecuInt.data()[0] == uInt(v));
  Vector<Int64> vecInt64 = vh.asArrayInt64();
  AlwaysAssertExit (vecInt64.size() == 1  &&  vecInt64.data()[0] == Int64(v));
  Vector<Float> vecfloat = vh.asArrayFloat();
  AlwaysAssertExit (vecfloat.size() == 1  &&  vecfloat.data()[0] == float(v));
  Vector<Double> vecdouble = vh.asArrayDouble();
  AlwaysAssertExit (vecdouble.size() == 1  &&  vecdouble.data()[0] == double(v));
  Vector<Complex> veccomplex = vh.asArrayComplex();
  AlwaysAssertExit (veccomplex.size() == 1  &&
                    veccomplex.data()[0] == Complex(float(v), 0));
  Vector<DComplex> vecdcomplex = vh.asArrayDComplex();
  AlwaysAssertExit (vecdcomplex.size() == 1  &&
                    vecdcomplex.data()[0] == DComplex(double(v), 0));
  Record rec;
  vh.toRecord (rec, "a");
  ValueHolder vhc;
  AlwaysAssertExit (vhc.isNull());
  vhc = ValueHolder::fromRecord (rec, "a");
  AlwaysAssertExit (!vhc.isNull());
  if (dt == TpUShort) {
    AlwaysAssertExit (vhc.dataType() == TpInt);
  } else {
    AlwaysAssertExit (vhc.dataType() == dt);
  }
  U vc;
  vhc.getValue (vc);
  AlwaysAssertExit (vc == U(v));
  cout << vh << ' ' << vhc << endl;
}

template<typename T, typename U> void doNeg(T v, U, DataType dt)
{
  ValueHolder vh(v);
  AlwaysAssertExit (!vh.isNull());
  AlwaysAssertExit (vh.dataType() == dt);
  AlwaysAssertExit (vh.asBool());
  AlwaysAssertExit (vh.asShort() == Short(v));
  AlwaysAssertExit (vh.asInt() == Int(v));
  AlwaysAssertExit (vh.asInt64() == Int64(v));
  AlwaysAssertExit (vh.asFloat() == float(v));
  AlwaysAssertExit (vh.asDouble() == double(v));
  AlwaysAssertExit (vh.asComplex() == Complex(float(v),0));
  AlwaysAssertExit (vh.asDComplex() == DComplex(double(v),0));
  Vector<Bool> vecbool = vh.asArrayBool();
  AlwaysAssertExit (vecbool.size() == 1  &&  vecbool.data()[0] == True);
  Vector<Short> vecShort = vh.asArrayShort();
  AlwaysAssertExit (vecShort.size() == 1  &&  vecShort.data()[0] == Short(v));
  Vector<Int> vecInt = vh.asArrayInt();
  AlwaysAssertExit (vecInt.size() == 1  &&  vecInt.data()[0] == Int(v));
  Vector<Int64> vecInt64 = vh.asArrayInt64();
  AlwaysAssertExit (vecInt64.size() == 1  &&  vecInt64.data()[0] == Int64(v));
  Vector<Float> vecfloat = vh.asArrayFloat();
  AlwaysAssertExit (vecfloat.size() == 1  &&  vecfloat.data()[0] == float(v));
  Vector<Double> vecdouble = vh.asArrayDouble();
  AlwaysAssertExit (vecdouble.size() == 1  &&  vecdouble.data()[0] == double(v));
  Vector<Complex> veccomplex = vh.asArrayComplex();
  AlwaysAssertExit (veccomplex.size() == 1  &&
                    veccomplex.data()[0] == Complex(float(v), 0));
  Vector<DComplex> vecdcomplex = vh.asArrayDComplex();
  AlwaysAssertExit (vecdcomplex.size() == 1  &&
                    vecdcomplex.data()[0] == DComplex(double(v), 0));
  Record rec;
  vh.toRecord (rec, "a");
  ValueHolder vhc;
  AlwaysAssertExit (vhc.isNull());
  vhc = ValueHolder::fromRecord (rec, "a");
  AlwaysAssertExit (!vhc.isNull());
  AlwaysAssertExit (vhc.dataType() == dt);
  U vc;
  vhc.getValue (vc);
  AlwaysAssertExit (vc == v);
  cout << vh << ' ' << vhc << endl;
}

void doComplex (const Complex& v)
{
  ValueHolder vh(v);
  AlwaysAssertExit (!vh.isNull());
  AlwaysAssertExit (vh.dataType() == TpComplex);
  AlwaysAssertExit (vh.asComplex() == v);
  AlwaysAssertExit (vh.asDComplex() == DComplex(v));
  Vector<Complex> vec = vh.asArrayComplex();
  AlwaysAssertExit (vec.size() == 1  &&  vec.data()[0] == v);
  Vector<DComplex> vecd = vh.asArrayDComplex();
  AlwaysAssertExit (vecd.size() == 1  &&  vecd.data()[0] == DComplex(v));
  Record rec;
  vh.toRecord (rec, "a");
  ValueHolder vhc;
  AlwaysAssertExit (vhc.isNull());
  vhc = ValueHolder::fromRecord (rec, "a");
  AlwaysAssertExit (!vhc.isNull());
  AlwaysAssertExit (vhc.dataType() == TpComplex);
  Complex vc;
  vhc.getValue (vc);
  AlwaysAssertExit (vc == v);
  cout << vh << ' ' << vhc << endl;
}

void doDComplex (const DComplex& v)
{
  ValueHolder vh(v);
  AlwaysAssertExit (!vh.isNull());
  AlwaysAssertExit (vh.dataType() == TpDComplex);
  AlwaysAssertExit (vh.asComplex() == Complex(v));
  AlwaysAssertExit (vh.asDComplex() == v);
  Vector<Complex> vec = vh.asArrayComplex();
  AlwaysAssertExit (vec.size() == 1  &&  vec.data()[0] == Complex(v));
  Vector<DComplex> vecd = vh.asArrayDComplex();
  AlwaysAssertExit (vecd.size() == 1  &&  vecd.data()[0] == v);
  Record rec;
  vh.toRecord (rec, "a");
  ValueHolder vhc;
  AlwaysAssertExit (vhc.isNull());
  vhc = ValueHolder::fromRecord (rec, "a");
  AlwaysAssertExit (!vhc.isNull());
  AlwaysAssertExit (vhc.dataType() == TpDComplex);
  DComplex vc;
  vhc.getValue (vc);
  AlwaysAssertExit (vc == v);
  cout << vh << ' ' << vhc << endl;
}

void doString (const String& v)
{
  ValueHolder vh1(v);
  AlwaysAssertExit (!vh1.isNull());
  AlwaysAssertExit (vh1.dataType() == TpString);
  AlwaysAssertExit (vh1.asString() == v);
  ValueHolder vh2(v.c_str());
  AlwaysAssertExit (!vh2.isNull());
  AlwaysAssertExit (vh2.dataType() == TpString);
  AlwaysAssertExit (vh2.asString() == v);
  Vector<String> vec = vh1.asArrayString();
  AlwaysAssertExit (vec.size() == 1  &&  vec.data()[0] == v);
  Record rec;
  vh1.toRecord (rec, "a");
  ValueHolder vhc;
  AlwaysAssertExit (vhc.isNull());
  vhc = ValueHolder::fromRecord (rec, "a");
  AlwaysAssertExit (!vhc.isNull());
  AlwaysAssertExit (vhc.dataType() == TpString);
  String vc;
  vhc.getValue (vc);
  AlwaysAssertExit (vc == v);
  ValueHolder vh3("123");
  ValueHolder vh4(vhc);
  vh3 = vhc;
  cout << vh1 << ' ' << vh2 << ' ' << vhc << ' ' << vh3 << ' ' << vh4 << endl;
}

void doArrayDComplex (const Array<DComplex>& v)
{
  ValueHolder vh(v);
  AlwaysAssertExit (!vh.isNull());
  AlwaysAssertExit (vh.dataType() == TpArrayDComplex);
  Array<Complex> ac(v.shape());
  convertArray (ac, v);
  AlwaysAssertExit (allEQ (vh.asArrayComplex(), ac));
  AlwaysAssertExit (allEQ (vh.asArrayDComplex(), v));
  Record rec;
  vh.toRecord (rec, "a");
  ValueHolder vhc;
  AlwaysAssertExit (vhc.isNull());
  vhc = ValueHolder::fromRecord (rec, "a");
  AlwaysAssertExit (!vhc.isNull());
  AlwaysAssertExit (vhc.dataType() == TpArrayDComplex);
  Array<DComplex> vc;
  vhc.getValue (vc);
  AlwaysAssertExit (allEQ (vc, v));
  cout << vh << ' ' << vhc << endl;
}

void doArrayString (const Array<String>& v)
{
  ValueHolder vh1(v);
  AlwaysAssertExit (!vh1.isNull());
  AlwaysAssertExit (vh1.dataType() == TpArrayString);
  AlwaysAssertExit (allEQ(vh1.asArrayString(), v));
  Record rec;
  vh1.toRecord (rec, "a");
  ValueHolder vhc;
  AlwaysAssertExit (vhc.isNull());
  vhc = ValueHolder::fromRecord (rec, "a");
  AlwaysAssertExit (!vhc.isNull());
  AlwaysAssertExit (vhc.dataType() == TpArrayString);
  Array<String> vc;
  vhc.getValue (vc);
  AlwaysAssertExit (allEQ(vc, v));
  cout << vh1 << ' ' << vhc << endl;
}

int main()
{
  try {
    doBool(True);
    doBool(False);
    doNeg(Short(-4), Short(0), TpShort);
    doNeg(Int(-7), Int(0), TpInt);
    doNeg(Int64(-40), Int64(0), TpInt64);
    doNeg(float(-4.1), float(0), TpFloat);
    doNeg(double(-4.7), double(0), TpDouble);
    doPos(uChar(13), uChar(0), TpUChar);
    doPos(Short(4), Short(0), TpShort);
    doPos(uShort(14), uShort(0), TpUShort);
    doPos(Int(17), Int(0), TpInt);
    doPos(uInt(10), uInt(0), TpUInt);
    doPos(Int64(40), Int64(0), TpInt64);
    doPos(float(4.1), Float(0), TpFloat);
    doPos(double(4.7), Double(0), TpDouble);
    doString(String());
    doString("");
    doString("astring");
    doComplex(Complex(3.1, -1.5));
    doDComplex(DComplex(-3.2, 11.5));

    Array<DComplex> dcomplexs(IPosition(2,1,2));
    dcomplexs(IPosition(2,0,0)) = DComplex(-1.311, 1.422);
    dcomplexs(IPosition(2,0,1)) = DComplex(-11.311, 21.422);
    doArrayDComplex (dcomplexs);

    Array<String> strings(IPosition(2,1,1));
    strings = "abc";
    doArrayString (strings);
  } catch (AipsError& x) {
    cout << "Caught an exception: " << x.getMesg() << endl;
    return 1;
  } 
  cout << "OK" << endl;
  return 0;                           // exit with success status
}
