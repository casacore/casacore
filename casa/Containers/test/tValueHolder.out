1
1 1
0
0 0
-4 -4
-7 -7
-40 -40
-4.1 -4.1
-4.7 -4.7
13 13
4 4
14 14
17 17
10 10
40 40
4.1 4.1
4.7 4.7
    
    
astring astring astring astring astring
(3.1,-1.5) (3.1,-1.5)
(-3.2,11.5) (-3.2,11.5)
Axis Lengths: [1, 2]  (NB: Matrix in Row/Column order)
[(-1.311,1.422), (-11.311,21.422)]
 Axis Lengths: [1, 2]  (NB: Matrix in Row/Column order)
[(-1.311,1.422), (-11.311,21.422)]

Axis Lengths: [1, 1]  (NB: Matrix in Row/Column order)
[abc]
 Axis Lengths: [1, 1]  (NB: Matrix in Row/Column order)
[abc]

OK
